sound/soc/sof/topology.c

```diff
@@ -1385,67 +1375,167 @@ static int sof_widget_load_siggen(struct snd_soc_component *scomp, int index,
+			       struct sof_ipc_comp_reply *r)
+
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
 	struct snd_soc_tplg_private *private = &tw->priv;
-	struct sof_ipc_comp_eq_iir *eq;
+	struct sof_ipc_comp_eq_fir *fir;
+	size_t ipc_size = 0, fir_data_size = 0;
+	void *fir_data_p;
 	int ret;
 
-	eq = kzalloc(sizeof(*eq), GFP_KERNEL);
-	if (!eq)
+	fir_data_size = le32_to_cpu(private->size) - SOF_EFFECT_DATA_SIZE;
+	ipc_size = sizeof(struct sof_ipc_comp_eq_fir) + fir_data_size;
+	fir_data_p = private->data + SOF_EFFECT_DATA_SIZE;
+
+	fir = kzalloc(ipc_size, GFP_KERNEL);
+	if (!fir)
 		return -ENOMEM;
 
-        /* configure IIR EQ IPC message */
-	eq->comp.hdr.size = sizeof(*eq);
-	eq->comp.hdr.cmd = SOF_IPC_GLB_TPLG_MSG | SOF_IPC_TPLG_COMP_NEW;
-	eq->comp.id = swidget->comp_id;
-	eq->comp.pipeline_id = index;
+	/* configure mixer IPC message */
+	fir->comp.hdr.size = ipc_size;
+	fir->comp.hdr.cmd = SOF_IPC_GLB_TPLG_MSG | SOF_IPC_TPLG_COMP_NEW;
+	fir->comp.id = swidget->comp_id;
+	fir->comp.type = SOF_COMP_EQ_FIR;
+	fir->comp.pipeline_id = index;
 
-	eq->comp.type = SOF_COMP_EQ_IIR;
-	ret = sof_parse_tokens(scomp, eq, eq_iir_tokens,
-			       ARRAY_SIZE(eq_iir_tokens), private->array,
-			       le32_to_cpu(private->size));
-	if (ret) {
-		dev_err(sdev->dev, "error: parse EQ tokens failed %d\n",
-			private->size);
-		goto err;
-	}
+	memcpy(&fir->data, fir_data_p, fir_data_size);
+
+	fir->size = fir_data_size;
 
-	ret = sof_parse_tokens(scomp, &eq->config, comp_tokens,
+	ret = sof_parse_tokens(scomp, &fir->config, comp_tokens,
 			       ARRAY_SIZE(comp_tokens), private->array,
-			       le32_to_cpu(private->size));
-	if (ret) {
-		dev_err(sdev->dev, "error: parse EQ.cfg tokens failed %d\n",
+			       SOF_EFFECT_DATA_SIZE);
+	if (ret != 0) {
+		dev_err(sdev->dev, "error: parse fir.cfg tokens failed %d\n",
 			le32_to_cpu(private->size));
 		goto err;
 	}
 
-	dev_dbg(sdev->dev, "eq iir %s created\n", swidget->widget->name);
+	sof_dbg_comp_config(scomp, &fir->config);
+
+	swidget->private = (void *)fir;
+
+	ret = sof_ipc_tx_message(sdev->ipc, fir->comp.hdr.cmd, fir,
+				 ipc_size, r, sizeof(*r));
+	if (ret >= 0)
+		return ret;
+err:
+	kfree(fir);
+	return ret;
+}
+
+static int sof_effect_iir_load(struct snd_soc_component *scomp, int index,
+			       struct snd_sof_widget *swidget,
+			       struct snd_soc_tplg_dapm_widget *tw,
+			       struct sof_ipc_comp_reply *r)
+
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_soc_tplg_private *private = &tw->priv;
+	struct sof_ipc_comp_eq_iir *iir;
+	size_t ipc_size = 0, iir_data_size = 0;
+	void *iir_data_p;
+	int ret;
+
+	iir_data_size = le32_to_cpu(private->size) - SOF_EFFECT_DATA_SIZE;
+	ipc_size = sizeof(struct sof_ipc_comp_eq_iir) + iir_data_size;
+	iir_data_p = private->data + SOF_EFFECT_DATA_SIZE;
+
+	iir = kzalloc(ipc_size, GFP_KERNEL);
+	if (!iir)
+		return -ENOMEM;
+
+	/* configure mixer IPC message */
```

Comment:
typo in the comment?

Reply:
yeah will fix

Reply:
btw this also in tonegen...where I copied this part :)